src/Microsoft.VisualStudio.Composition/ReflectionHelpers.cs

```diff
@@ -677,6 +691,40 @@ internal static int GetMetadataTokenSafe(this MemberInfo memberInfo)
             return memberInfo.MetadataToken;
         }
 
+        internal static bool TryGetContainerType(Type type, out TypeInfo containerType)
+        {
+            Requires.NotNull(type, nameof(type));
+
+            // Discover the ICollection<T> or ICollection<Lazy<T, TMetadata>> interface implemented by this type.
+            var icollectionTypes =
```

Comment:
This is a non-trivial bit of code that you seem to have copied from `PartDiscovery.GetElementTypeFromMany`. Can you share the code instead? That way perf and functional changes will be consistent between the two.